api/src/syscall/mm/mincore.rs

```diff
@@ -0,0 +1,152 @@
+/// * `Err(EINVAL)` - addr is not a multiple of the page size
+/// * `Err(ENOMEM)` - length is greater than (TASK_SIZE - addr), or negative length,
+///                   or addr to addr+length contained unmapped memory
+///
+/// # Notes from Linux man page
+/// - The least significant bit (bit 0) is set if page is resident in memory
+/// - Bits 1-7 are reserved and currently cleared
+/// - Information is only a snapshot; pages can be swapped at any moment
+///
+/// # Linux Errors
+/// - EAGAIN:  kernel temporarily out of resources
+/// - EFAULT: vec points to invalid address
+/// - EINVAL: addr not page-aligned
+/// - ENOMEM: length > (TASK_SIZE - addr), negative length, or unmapped memory
+pub fn sys_mincore(addr:  usize, length: usize, vec: *mut u8) -> AxResult<isize> {
+    debug!("sys_mincore <= addr: {addr:#x}, length: {length:#x}, vec: {vec:?}");
+
+    // EINVAL: addr must be a multiple of the page size
+    // TODO: Arceos should support unified PAGE_SIZE constant
+    if !addr.is_multiple_of(PAGE_SIZE_4K) {
+        return Err(AxError::InvalidInput);
+    }
+
+    // EFAULT: vec must not be null (basic check, vm_write_slice will do full validation)
+    if vec.is_null() {
+        return Err(AxError::BadAddress);
+    }
+
+    // Special case: length 0 is valid and returns immediately
+    if length == 0 {
+        return Ok(0);
+    }
+
+    let start_addr = VirtAddr::from(addr);
+
+    // ENOMEM: Check for overflow (simulates "length > TASK_SIZE - addr")
+    // This catches negative lengths interpreted as large unsigned values
+    start_addr.checked_add(length).ok_or_else(|| {
+        LinuxError::ENOMEM
+    })?;
+
+    // Calculate number of pages to check
+    // Formula from man page: (length + PAGE_SIZE - 1) / PAGE_SIZE
+    let page_count = (length + PAGE_SIZE_4K - 1) / PAGE_SIZE_4K;
+
+    // Get current address space
+    let curr = current();
+    let aspace = curr.as_thread().proc_data.aspace.lock();
+
+    // Allocate temporary buffer for results
+    // Initialize all bytes to 0 (non-resident, all reserved bits clear)
+    let mut result_vec = vec![0u8; page_count];
+
+    // Check each page in the range [addr, addr+length)
+    let mut current_page = start_addr. align_down_4k();
+
+    for i in 0..page_count {
+        // ENOMEM: Check if this page is within a valid VMA (Virtual Memory Area)
+        // Linux returns ENOMEM for "unmapped memory"
+        aspace.find_area(current_page).ok_or_else(|| {
+            // This address is not mapped - return ENOMEM per Linux spec
+            LinuxError:: ENOMEM
+        })?;
+
+        // Verify we have at least USER access permission
+        // (Though find_area likely already ensures this for user mappings)
+        if !aspace.can_access_range(current_page, PAGE_SIZE_4K, MappingFlags::USER) {
+            // Mapped but not accessible - treat as ENOMEM per Linux behavior
+            drop(aspace);
+            return Err(LinuxError::ENOMEM. into());
+        }
```

Comment:
These two steps can be simplified: using the area info returned by `find_area` you can directly check user permission.